samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java

```diff
@@ -29,60 +37,136 @@
+   *
+   * @param nextOperator  the next operator to propagate results to
    */
   void registerNextOperator(OperatorImpl<RM, ?> nextOperator) {
-    nextOperators.add(nextOperator);
+    if (!initialized) {
+      throw new IllegalStateException(
+          String.format("Attempted to register next operator before initializing operator %s.", getOpName()));
+    }
+    this.registeredOperators.add(nextOperator);
   }
 
   /**
-   * Perform the transformation required for this operator and call the downstream operators.
+   * Handle the incoming {@code message} for this {@link OperatorImpl} and propagate results to registered operators.
+   * <p>
+   * Delegates to {@link #handleMessage(Object, MessageCollector, TaskCoordinator)} for handling the message.
    *
-   * Must call {@link #propagateResult} to propagate the output to registered downstream operators correctly.
+   * @param message  the input message
+   * @param collector  the {@link MessageCollector} for this message
+   * @param coordinator  the {@link TaskCoordinator} for this message
+   */
+  public final void onMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    this.messageCounter.inc();
+    long startNs = System.nanoTime();
+    Collection<RM> results = handleMessage(message, collector, coordinator);
+    long endNs = System.nanoTime();
+    this.handleMessageTimer.update(endNs - startNs);
+
+    results.forEach(rm ->
+        this.registeredOperators.forEach(op ->
+            op.onMessage(rm, collector, coordinator)));
+  }
+
+  /**
+   * Handle the incoming {@code message} and return the results to be propagated to registered operators.
    *
    * @param message  the input message
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
+   * @return  results of the transformation
    */
-  public abstract void onNext(M message, MessageCollector collector, TaskCoordinator coordinator);
+  protected abstract Collection<RM> handleMessage(M message, MessageCollector collector,
+      TaskCoordinator coordinator);
 
   /**
-   * Invoked at every tick. This method delegates to {@link #onTimer(MessageCollector, TaskCoordinator)}
+   * Handle timer ticks for this {@link OperatorImpl} and propagate the results and timer tick to registered operators.
+   * <p>
+   * Delegates to {@link #handleTimer(MessageCollector, TaskCoordinator)} for handling the timer tick.
    *
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
    */
-  public final void onTick(MessageCollector collector, TaskCoordinator coordinator) {
-    onTimer(collector, coordinator);
-    nextOperators.forEach(sub -> sub.onTick(collector, coordinator));
+  public final void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long startNs = System.nanoTime();
```

Comment:
IIUC, We have a mode in `HDFS` to disable timers entirely. That way, we don't make a lot of calls to `System.nanoTime()`. 

Would be nice to support that as a part of a different PR.